secant/Utils/Zatoshi.swift

```diff
@@ -0,0 +1,68 @@
+//
+//  Zatoshi.swift
+//  secant-testnet
+//
+//  Created by Lukáš Korba on 26.05.2022.
+//
+
+import Foundation
+
+struct Zatoshi {
+    enum Constants {
+        static let oneZecInZatoshi: Int64 = 100_000_000
+        static let maxZecs: Int64 = 21_000_000
+        static let maxZatoshi: Int64 = Constants.oneZecInZatoshi * Constants.maxZecs
+    }
+    
+    static var nsDecimalHandler = NSDecimalNumberHandler(
+        roundingMode: NSDecimalNumber.RoundingMode.bankers,
+        scale: 8,
+        raiseOnExactness: true,
+        raiseOnOverflow: true,
+        raiseOnUnderflow: true,
+        raiseOnDivideByZero: true
+    )
+    
+    @Clamped(-Constants.maxZatoshi...Constants.maxZatoshi)
+    var amount: Int64 = 0
+
+    /// Converts `Zatoshi` to `NSDecimalNumber`
+    var asZec: NSDecimalNumber {
+        NSDecimalNumber(decimal: Decimal(amount) / Decimal(Constants.oneZecInZatoshi))
+    }
+
+    /// Converts `Zatoshi` to human readable format, up to 8 fraction digits
+    var asHumanReadableZec: String {
+        asZec.rounding(accordingToBehavior: Zatoshi.nsDecimalHandler).stringValue
+    }
+
+    /// Converts `Decimal` to `Zatoshi`
+    static func from(zec: Decimal) -> Zatoshi {
+        let zec2zatoshi = Decimal(Constants.oneZecInZatoshi) * zec
+        return Zatoshi(amount: NSDecimalNumber(decimal: zec2zatoshi).int64Value)
+    }
+
+    /// Converts `String` to `Zatoshi`
+    static func from(zec: String, formatter: NumberFormatter) -> Zatoshi? {
```

Comment:
Who creates the `NumberFormatter`?  It seems like clients should only have a little bit of influence on the parsing.

On the Android side, we don't give clients full control over the formatting but we did need a way to cache the formatting objects a bit to ensure that multiple method calls could use the same objects.  (A concern is that system locale can change at any time, so we don't want to keep asking for locale when making several API calls in a row but rather ask once during a composition of the UI).

Reply:
The reducer (TCA). This function is pure and independent, the formatter is going to be injected (#312)

Reply:
Imagine this code ends up moving into the SDK (or a helper library of the SDK).

Would you want a wrapper class so that only the SDK can generate the number formatter?  It can still be constructed by the UI and injected in, but this ensures that clients of the API can't create arbitrary number formatters.

e.g. this sort of pseudo-swift example
```
struct ZecFormatter {
    let numberFormatter: NumberFormatter
    constructor ZecFormatter() {
           var formatter = NumberFormatter()
            formatter.maximumFractionDigits = 8
            formatter.maximumIntegerDigits = 8
            formatter.numberStyle = .decimal
            formatter.usesGroupingSeparator = true
        
            numberFormatter = formatter
    }
}
```

Reply:
@LukasKorba Is this fully resolved?

We will be moving Zatoshi into the SDK for sure in the future, since it is a key public API.

We may or may not move the Zec formatting.  (Zec formatting could be turned into extension functions on Zatoshi that live in the app if we'd like).  But if we do move the Zec formatting into the SDK, I feel like the SDK should have a mechanism to control generation of the formatter.  So this is one item for consideration.  We can defer it until #312 but I don't think NumberFormatter is the right API for `static func from(zec: String, formatter: NumberFormatter)`

Reply:
This is a great point. 

Maybe if we move this to the SDK we can use a default value for the formatter parameter.

I think that most of the multi coin wallets we want to aim for as potential users must have their own number formatters

Reply:
As long as we want to create an API that is flexible enough to inject formatter, the location doesn't matter. Either in the wallet or SDK. For either of those we will create a WrappedNumberFormatter and inject it when using the API `static func from(zec: String, formatter: NumberFormatter)`. So the SDK can be in control of let's say default SDK formatter unless anybody want's to override it.

Inflexible solution would implement number formatter being part of the Zatoshi type and never let anybody to control it. I hope it answered your point, otherwise I'm not sure if I understand the concern.

Reply:
> As long as we want to create an API that is flexible enough to inject formatter, the location doesn't matter. Either in the wallet or SDK. For either of those we will create a WrappedNumberFormatter and inject it when using the API `static func from(zec: String, formatter: NumberFormatter)`. So the SDK can be in control of let's say default SDK formatter unless anybody want's to override it.
> 
> Inflexible solution would implement number formatter being part of the Zatoshi type and never let anybody to control it. I hope it answered your point, otherwise I'm not sure if I understand the concern.

I'm ok with this. I think the idea is to make a clear statement about what the default is and whomever uses the injection does it at their own risk :)

Reply:
Yes, Pacu has it.  I just want it to be super clear what we typically expect clients to pass in here and deviating from that is at their own risk.

Reply:
@LukasKorba let's try to add the default value and make the API really clear. 

Excellent work gentlemen

Reply:
I think we are all on the same boat. My plan was to:
1. Implement our (Zcash) formatter and let it be in the SDK - the default one
2. Move Zatoshi to the SDK as well
3. Create WrappedNumberFormatter, its live instance is going to use the one from the SDK
4. TCA will inject WrappedNumberFormatter dependency (live one)

The wallet will never define another NumberFormatter, users of SDK can do whatever they want